cli: add config command and config validate subcommand to nomad CLI

th0m opened the pull request:
This command is similar to the `consul validate` command and can be used to perform configuration validation before actually restarting Nomad and risking a failure if there is an issue in the configuration.

Also added three test cases to the `command` test suite:

1. Empty directory which is acceptable
1. Working configuration
1. Non-working configuration

jrasell reviewed (commented):
Thanks for much for this @th0m.

I would be curious of your thoughts, as well as others, around the command `nomad config validate` rather than `nomad config-validate`? In my mind, this allows for additional config based sub-commands in the future thus providing future flexibility while keeping the experience consistent.

th0m commented:
Makes sense to me @jrasell, happy to change it.

jrasell commented on command/config_validate.go:
```diff
@@ -0,0 +1,73 @@
+
+	for _, path := range configPath {
+		current, err := agent.LoadConfig(path)
+		if err != nil {
+			c.Ui.Error(fmt.Sprintf(
+				"Error loading configuration from %s: %s", path, err))
+			return 1
```
I am curious on your thoughts here around the use of [multierror](https://github.com/hashicorp/go-multierror)? In the event of multiple files having validation issues, the command can output all the errors in a single run, rather than needing the operator to run the command multiple times to fix each issue which is displayed.

th0m replied:
Great idea, I'll use multierror.

jrasell reviewed (changes requested):
Thanks for the update. I have suggested a couple of changes along with a question regarding the error handling which I would be keen to get your feedback on as a target user for this command.

dadgar commented:
@jrasell @th0m It may be good to refactor the [isValidConfig](https://github.com/hashicorp/nomad/blob/master/command/agent/command.go#L290) call so that it can be used here. Currently, the command just validates the config can be parsed, but it may have settings that will cause Nomad not to be able to start:

th0m commented:
Thanks for the review @dadgar and @jrasell.
I just rebased my branch and will follow up on your comments shortly.

th0m commented:
@dadgar following your suggestion I've made the `isValidConfig` function visible to other packages to be able to use it here and perform config validation in addition to parsability validation.
I've added a test that catches a bad configuration that is otherwise parsable.

@jrasell I think I have addressed all your comments, can you please re-review?

Thanks both.

th0m commented:
@jrasell sorry about the repeated ping but in case you missed my previous message this PR is now ready for review :)
Thank you!

th0m commented:
@jrasell @tgross may I have a review here when you have a minute? Thank you so much!

tgross commented on command/config_validate.go:
```diff
@@ -0,0 +1,85 @@
+	helpText := `
+Usage: nomad config validate <config_path> [<config_path...>]
+
+  Performs a thorough sanity test on Nomad configuration files. For each file
+  or directory given, the validate command will attempt to parse the contents
+  just as the "nomad agent" command would, and catch any errors.
+
+  This is useful to do a test of the configuration only, without actually
+  starting the agent. This performs all of the validation the agent would, so
+  this should be given the complete set of configuration files that are going
+  to be loaded by the agent. This command cannot operate on partial
+  configuration fragments since those won't pass the full agent validation.
+
+  Returns 0 if the configuration is valid, or 1 if there are problems.
```
Let's tighten up the text here for inclusive language, for grammar, to match some existing `nomad agent -config` language, and to remove some redundant phrasing:

```suggestion
  Perform validation on a set of Nomad configuration files. This is useful
  to test the Nomad configuration without starting the agent.

  Accepts the path to either a single config file or a directory of
  config files to use for configuring the Nomad agent. This option may
  be specified multiple times. If multiple config files are used, the
  values from each will be merged together. During merging, values from
  files found later in the list are merged over values from previously
  parsed files.

  This command cannot operate on partial configuration fragments since
  those won't pass the full agent validation. This command does not
  require an ACL token.

  Returns 0 if the configuration is valid, or 1 if there are problems.
```

tgross commented on command/config_validate.go:
```diff
@@ -0,0 +1,85 @@
+	config := agent.DefaultConfig()
+
+	for _, path := range configPath {
+		fc, err := agent.LoadConfig(path)
+		if fc == nil || reflect.DeepEqual(fc, &agent.Config{}) {
+			c.Ui.Warn(fmt.Sprintf("No configuration loaded from %s", path))
+		}
+		if err != nil {
+			multierror.Append(&mErr, fmt.Errorf(
+				"Error loading configuration from %s: %s", path, err))
+			continue
+		}
```
It's probably safe here, but in general it's better to check the error first and then check the result. Let's swap these two blocks:

```suggestion
		fc, err := agent.LoadConfig(path)
		if err != nil {
			multierror.Append(&mErr, fmt.Errorf(
				"Error loading configuration from %s: %s", path, err))
			continue
		}
		if fc == nil || reflect.DeepEqual(fc, &agent.Config{}) {
			c.Ui.Warn(fmt.Sprintf("No configuration loaded from %s", path))
		}
```

tgross reviewed (changes requested):
Looking pretty good @th0m! I've left some comments but mostly it's just tightening up language at this point.

th0m commented:
Thank you @tgross, I have committed both of your suggestions and squashed the three commits into one.

tgross reviewed (approved):
LGTM

th0m commented:
Thanks @tgross!
